noir-projects/noir-contracts/contracts/inclusion_proofs_contract/src/test.nr

```diff
@@ -0,0 +1,366 @@
+    env.advance_block_by(1);
+
+    // Deploy contract and initialize
+    let initializer = InclusionProofs::interface().constructor(initial_value);
+    let inclusion_proofs_contract =
+        env.deploy_self("InclusionProofs").with_public_void_initializer(initializer);
+    let contract_address = inclusion_proofs_contract.to_address();
+    env.advance_block_by(1);
+    (&mut env, contract_address, owner)
+}
+
+#[test]
+unconstrained fn note_flow() {
+    // This test creates a note and checks certain properties about it, namely its inclusion in the trees,
+    // the non-inclusion of its nullifier, and its validity (the combination of the previous two)
+    let (env, contract_address, owner) = setup(INITIAL_VALUE);
+    env.impersonate(owner);
+
+    let NOTE_VALUE = 69;
+    InclusionProofs::at(contract_address).create_note(owner, NOTE_VALUE).call(&mut env.private());
+
+    env.advance_block_by(1);
+
+    let note_creation_block_number = env.block_number();
+
+    // We advance by another block to make sure that the note creation block number != our current block number
+    env.advance_block_by(1);
+
+    let current_contract_address = get_contract_address();
+    cheatcodes::set_contract_address(contract_address);
+
+    // We fetch the note we created and make sure that it is valid.
+    let note = InclusionProofs::get_note(owner);
+    cheatcodes::set_contract_address(current_contract_address);
+
+    assert(note.owner.eq(owner));
+    assert(note.value.eq(NOTE_VALUE));
+
+    // Each of these tests (note inclusion, note non-nullification, and validity (inclusion & non-nullification)) check the assertion at the block of creation of note, as well as at the "current" block
+    InclusionProofs::at(contract_address)
+        .test_note_inclusion(owner, true, note_creation_block_number, false)
+        .call(&mut env.private());
+    InclusionProofs::at(contract_address).test_note_inclusion(owner, false, 0, false).call(
+        &mut env.private(),
+    );
+
+    InclusionProofs::at(contract_address)
+        .test_note_not_nullified(owner, true, note_creation_block_number, false)
+        .call(&mut env.private());
+    InclusionProofs::at(contract_address).test_note_not_nullified(owner, false, 0, false).call(
+        &mut env.private(),
+    );
+
+    InclusionProofs::at(contract_address)
+        .test_note_validity(owner, true, note_creation_block_number, false)
+        .call(&mut env.private());
+    InclusionProofs::at(contract_address).test_note_validity(owner, false, 0, false).call(
+        &mut env.private(),
+    );
+}
+
+#[test]
+unconstrained fn nullify_note_flow() {
+    // This test creates a note, nullifies it, and checks certain properties about it, namely its inclusion in the trees,
+    // the non-inclusion of its nullifier, and its validity (the combination of the previous two). These properties are checked before nullification.
```

Comment:
This comment is weird. It says that you nullify the note and then you prove that it's not nullified.